Emacs text registers

justinhopkins opened the pull request:
Greetings whitphx!

First of all, thank you for maintaining awesome-emacs-mcx! It is unbelievably excellent for emacs nerds like myself.

I've been experimenting with implementing Emacs text registers. "Text registers" are effectively multiple clipboards that you can identify with a single keystroke. Full-blown emacs registers can do a lot, but so far I'm just trying to implement two commands - "register save" and "register insert". 

This implementation seems to play nicely with the C-x r commands that work with rectangles.

There are some extensive changes, so I expect you may object to some design decisions. I'm happy to make changes based on your feedback.

Hope to hear from you


Reference:
https://www.gnu.org/software/emacs/manual/html_node/emacs/Text-Registers.html

^ From the documentation:

    register save:
    C-x r s r
    Copy region into register r (copy-to-register).

    register insert:
    C-x r i r
    Insert text from register r (insert-register).

whitphx commented on src/emulator.ts:
```diff
@@ -541,4 +564,61 @@ export class EmacsEmulator implements IEmacsController, vscode.Disposable {
     });
   }
+
+  public saveRegister(arg: string): void {
```
It would be great if you write tests for the new commands.
Just leave it if it's hard, as I will make some after merging the PR.

justinhopkins replied:
Understood. Please have a look at how I've modified EmulatorMap and EmacsEmulator to have a textRegister Map(). If this approach looks good, I'll start writing unit tests

whitphx commented on src/emulator.ts:
```diff
@@ -541,4 +564,61 @@ export class EmacsEmulator implements IEmacsController, vscode.Disposable {
+    let i = 0;
+    let combinedtext = "";
+    while (i < selections.length) {
+      combinedtext = combinedtext + this.textEditor.document.getText(selections[i]);
```
While it's not needed in this PR, the copy/paste functions of the kill/yank commands  should be shared with these methods. They are compatible with multi-cursors.

justinhopkins replied:
A good point. Still researching

whitphx reviewed (commented):
Thank you very much for your contribution!
Let me point out some things as comments.

Also, CI workflows have not been dispatched maybe because of misconfiguration. I updated the CI definition file on the main branch, so please update this branch with it and see if the CI is triggered.

justinhopkins commented on src/emulator.ts:
```diff
@@ -36,8 +38,25 @@ export interface IEmacsController {
   registerDisposable: (disposable: vscode.Disposable) => void;
 }
 
+export class TextRegister {
```
Agreed. I've updated it to just use a Map<string, string>

justinhopkins commented on src/emulator.ts:
```diff
@@ -541,4 +564,61 @@ export class EmacsEmulator implements IEmacsController, vscode.Disposable {
+      i++;
+    }
+    const msg = "Running RegisterSaveCommand to " + arg + " with this selection: " + combinedtext;
+    vscode.window.showInformationMessage(msg);
```
Roger that; changed it

justinhopkins commented:
> Thank you very much for your contribution! Let me point out some things as comments.
> 
> Also, CI workflows have not been dispatched maybe because of misconfiguration. I updated the CI definition file on the main branch, so please update this branch with it and see if the CI is triggered.

Re-merged with your changes on 'main' for changes from the .github/workflows/main.yaml.

Do you need me to do anything else for this PR? Would be happy to make changes if I've missed anything.

I'm also kind of waiting for your thoughts about how I've added textRegister as an object passed into the EmacsEmulator. If it looks alright, I'll go ahead and write some test cases for it.

whitphx commented:
@justinhopkins Thank you so much for your contributions and sorry for my late reply.
The current implementation looks nice and I think it's ready to write the tests.
Should I leave this PR open until you write the tests, or merge it first?

whitphx commented on src/commands/registers.ts:
```diff
@@ -0,0 +1,56 @@
+
+  private acceptingRegisterSaveCommand = false;
+
+  public startRegisterSaveCommand(): void {
```
Can you make this private?

whitphx replied:
Sorry this was originally my bad.
Fixed the original code you referred to as the basis: https://github.com/whitphx/vscode-emacs-mcx/pull/1680

whitphx commented on src/commands/registers.ts:
```diff
@@ -0,0 +1,56 @@
+
+  private acceptingRegisterInsertCommand = false;
+
+  public startRegisterInsertCommand(): void {
```
Can you make this private?

justinhopkins replied:
Yep. Changed

whitphx commented on src/emulator.ts:
```diff
@@ -592,4 +600,64 @@ export class EmacsEmulator implements IEmacsController, vscode.Disposable {
+    // Looking for how to insert-replace with selections highlighted.... must copy-paste from Yank command
+    const selections = this.textEditor.selections;
+
+    this.textEditor.edit((editBuilder) => {
```
```suggestion
    await this.textEditor.edit((editBuilder) => {
```

justinhopkins replied:
Changed

whitphx commented on src/emulator.ts:
```diff
@@ -592,4 +600,64 @@ export class EmacsEmulator implements IEmacsController, vscode.Disposable {
+    }
+
+    const msg = "Copying selection to register " + arg;
+    MessageManager.showMessage(msg);
```
Looks like the orignal Emacs doesn't show this message, so we shouldn't do it too?

justinhopkins replied:
Agreed. I've removed both the copying to register message and the inserting from register message.

whitphx reviewed (commented):
Let me suggest a few things ;)

justinhopkins commented:
> @justinhopkins Thank you so much for your contributions and sorry for my late reply. The current implementation looks nice and I think it's ready to write the tests. Should I leave this PR open until you write the tests, or merge it first?

No worries! Just made changes based on your feedback. Please merge it first, and I'll follow-up with unit tests!
Thanks!

whitphx commented:
OK, thanks!

whitphx commented:
Released in v0.50.0 👍
